askama_parser/src/lib.rs

```diff
@@ -155,6 +129,72 @@ impl fmt::Display for ParseError {
 pub(crate) type ParseErr<'a> = nom::Err<ErrorContext<'a>>;
 pub(crate) type ParseResult<'a, T = &'a str> = Result<(&'a str, T), ParseErr<'a>>;
 
+pub struct ErrorInfo {
```

Comment:
Although the direction is good, this still has too much going on. Suggest a few steps, maybe in a separate PR?

* Extract all of this code from `Ast::from_str()` into a `ParseError::new()` function
* Unify the row/column display to `{row}:{column}` for both path and non-path errors
* Extract an `ErrorInfo` type with an `impl Display` that takes care of the final formatting step, which will need to include a `path: Option<Rc<Path>>`
* To the extent that you need what is here `generate_row_and_column()` and `generate_error_info()` later in the PR, they should perhaps be constructors on `ErrorInfo`.

Also please keep functions in top-down order -- maybe do more self-review to check for style issues? Reviewing your changes is taking many round trips.

Reply:
Feedback is appreciated. I don't plan to send another PR for this, all changes are tied together and at this point your suggestions (even if great) are only improvements over this code. I'll implement what you suggested though.

> Also please keep functions in top-down order -- maybe do more self-review to check for style issues? Reviewing your changes is taking many round trips.

Well, as you said, it's code styling, so we're entering personal preferences. I don't mind following a project code's style but if there is no test to enforce it and tell me what and where I did something wrong, then you shouldn't expect people to remember it or to implement it as such. Therefore, I don't plan to try to remember what style you want to apply until there is a check for it, I work on way too many different codebases to have motivation for this. If this is a blocker for you, then I'll just fork askama from this point and implement what I need on my side while backporting changes from your repository to mine. Just to be clear: this is not a hostile take or anything, I really love `askama` and really enjoy your suggestions. It's just that you have high expectations for PRs, and I don't have the tools nor the time to live up to your expectations. So I'll let it up to you. I don't mind if you send commits to my PRs to fix code formatting, this is very welcome as well.